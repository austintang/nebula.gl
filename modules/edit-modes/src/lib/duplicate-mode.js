// @flow

import type { StartDraggingEvent, ModeProps } from '../types.js';
import type { FeatureCollection } from '../geojson-types.js';
import { TranslateMode } from './translate-mode.js';

export class DuplicateMode extends TranslateMode {
  handleStartDragging(event: StartDraggingEvent, props: ModeProps<FeatureCollection>) {
    super.handleStartDragging(event, props);

    if (this._geometryBeforeTranslate) {
      props.onEdit(this.getAddManyFeaturesAction(this._geometryBeforeTranslate, props.data));
    }
  }

  updateCursor(props: ModeProps<FeatureCollection>): boolean {
    if (this._isTranslatable) {
      props.onUpdateCursor('copy');
      return true;
    }
    props.onUpdateCursor(null);
    return false;
  }
}
