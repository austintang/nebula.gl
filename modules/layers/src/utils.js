// @flow

import destination from '@turf/destination';
import bearing from '@turf/bearing';
import pointToLineDistance from '@turf/point-to-line-distance';
import { point } from '@turf/helpers';
import type { Position, LineString } from './geojson-types.js';

export function toDeckColor(
  color?: ?[number, number, number, number],
  defaultColor: [number, number, number, number] = [255, 0, 0, 255]
): [number, number, number, number] {
  if (!Array.isArray(color)) {
    return defaultColor;
  }
  return [color[0] * 255, color[1] * 255, color[2] * 255, color[3] * 255];
}

//
// a GeoJSON helper function that calls the provided function with
// an argument that is the most deeply-nested array having elements
// that are arrays of primitives as an argument, e.g.
//
// {
//   "type": "MultiPolygon",
//   "coordinates": [
//       [
//           [[30, 20], [45, 40], [10, 40], [30, 20]]
//       ],
//       [
//           [[15, 5], [40, 10], [10, 20], [5, 10], [15, 5]]
//       ]
//   ]
// }
//
// the function would be called on:
//
// [[30, 20], [45, 40], [10, 40], [30, 20]]
//
// and
//
// [[15, 5], [40, 10], [10, 20], [5, 10], [15, 5]]
//
export function recursivelyTraverseNestedArrays(
  array: Array<any>,
  prefix: Array<number>,
  fn: Function
) {
  if (!Array.isArray(array[0])) {
    return true;
  }
  for (let i = 0; i < array.length; i++) {
    if (recursivelyTraverseNestedArrays(array[i], [...prefix, i], fn)) {
      fn(array, prefix);
      break;
    }
  }
  return false;
}

export function generatePointsParallelToLinePoints(
  p1: Position,
  p2: Position,
  groundCoords: Position
): Position[] {
  const lineString: LineString = {
    type: 'LineString',
    coordinates: [p1, p2]
  };
  const pt = point(groundCoords);
  const ddistance = pointToLineDistance(pt, lineString);
  const lineBearing = bearing(p1, p2);

  // Check if current point is to the left or right of line
  // Line from A=(x1,y1) to B=(x2,y2) a point P=(x,y)
  // then (x−x1)(y2−y1)−(y−y1)(x2−x1)
  const isPointToLeftOfLine =
    (groundCoords[0] - p1[0]) * (p2[1] - p1[1]) - (groundCoords[1] - p1[1]) * (p2[0] - p1[0]);

  // Bearing to draw perpendicular to the line string
  const orthogonalBearing = isPointToLeftOfLine < 0 ? lineBearing - 90 : lineBearing - 270;

  // Get coordinates for the point p3 and p4 which are perpendicular to the lineString
  // Add the distance as the current position moves away from the lineString
  const p3 = destination(p2, ddistance, orthogonalBearing);
  const p4 = destination(p1, ddistance, orthogonalBearing);

  return [p3.geometry.coordinates, p4.geometry.coordinates];
}

export const SortType = {
  MIN: 'min',
  MAX: 'max'
};

export class SortedList {
  _array: any[];
  _arraySize: number;
  _getValueToCompareFn: Function;
  _sortType: string;

  constructor(options: Object) {
    const { arraySize, getValueToCompareFn, sortType } = options;
    this._array = [];
    this._arraySize = arraySize;
    this._getValueToCompareFn = getValueToCompareFn;
    this._sortType = sortType;
  }

  push(value: any): boolean {
    let isArrayModified = false;
    const pushIndex = this._array.length;

    if (pushIndex < this._arraySize) {
      // Initial array population
      this._array[pushIndex] = value;
      isArrayModified = true;
    } else {
      // Value comparison function that accounts for the sortType
      const contextualValueComparison = (a, b) => (this._sortType === 'min' ? a < b : a > b);
      // If the value passed in this method will be sorted before the last item
      // of the array, replace the last item of the array with the value passed.
      const isPushable = contextualValueComparison(
        this._getValueToCompareFn(value),
        this._getValueToCompareFn(this._array[this._arraySize - 1])
      );
      if (isPushable) {
        this._array[this._arraySize - 1] = value;
        isArrayModified = true;
      }
    }

    if (this._arraySize > 1 && isArrayModified) {
      // Sort the array only if the array was modified. Sort the array so that
      // the last item in the array is ready to be replaced (i.e. it has the
      // greatest value if this.sortType === min, or has the smallest value if
      // this.sortType === max).
      const isSortMin = this._sortType === 'min';
      const contextualSortFunction = (a, b) =>
        this._getValueToCompareFn(isSortMin ? a : b) - this._getValueToCompareFn(isSortMin ? b : a);

      this._array.sort(contextualSortFunction);
    }

    return isArrayModified;
  }

  fromArray(valueArray: any[]) {
    valueArray.forEach(value => this.push(value));
  }

  toArray(): any[] {
    return this._array;
  }

  size(): number {
    return this._array.length;
  }
}
