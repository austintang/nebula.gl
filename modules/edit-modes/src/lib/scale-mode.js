// @flow

import turfBuffer from '@turf/buffer';
import bbox from '@turf/bbox';
import turfCentroid from '@turf/centroid';
import turfBearing from '@turf/bearing';
import bboxPolygon from '@turf/bbox-polygon';
import { point, featureCollection } from '@turf/helpers';
import { coordEach } from '@turf/meta';
import turfDistance from '@turf/distance';
import turfTransformScale from '@turf/transform-scale';
import { getCoord, getGeom } from '@turf/invariant';
import type { FeatureCollection, Position } from '../geojson-types.js';
import type {
  ModeProps,
  PointerMoveEvent,
  StartDraggingEvent,
  StopDraggingEvent,
  DraggingEvent,
  EditHandleFeature
} from '../types.js';
import { getPickedEditHandle } from '../utils';
import { BaseGeoJsonEditMode } from './geojson-edit-mode.js';
import { ImmutableFeatureCollection } from './immutable-feature-collection.js';

export class ScaleMode extends BaseGeoJsonEditMode {
  _selectedGeometry: ?FeatureCollection;
  _selectedEditHandle: ?EditHandleFeature;
  _cornerGuidePoints: Array<EditHandleFeature>;
  _previousMouseCoords: ?Position;
  _cursor: ?string;
  _isScaling: boolean = false;
  _isCompositionMode: boolean;

  constructor(options: ?{ isCompositionMode: boolean }) {
    super();
    const { isCompositionMode } = options || {};
    this._isCompositionMode = isCompositionMode || false;
  }

  _isSinglePointGeometrySelected = (): boolean => {
    const { features } = this._selectedGeometry || {};
    if (Array.isArray(features) && features.length === 1) {
      const { type } = getGeom(features[0]);
      return type === 'Point';
    }
    return false;
  };

  _getOppositeScaleHandle = (selectedHandle: EditHandleFeature) => {
    const selectedHandleIndex =
      selectedHandle && selectedHandle.properties && selectedHandle.properties.index;
    if (typeof selectedHandleIndex !== 'number') {
      return null;
    }
    const guidePointCount = this._cornerGuidePoints.length;
    const oppositeIndex = (selectedHandleIndex + guidePointCount / 2) % guidePointCount;
    return this._cornerGuidePoints.find(p => p.properties.index === oppositeIndex);
  };

  _getUpdatedData = (props: ModeProps<FeatureCollection>, editedData: FeatureCollection) => {
    let updatedData = new ImmutableFeatureCollection(props.data);
    const selectedIndexes = props.selectedIndexes;
    for (let i = 0; i < selectedIndexes.length; i++) {
      const selectedIndex = selectedIndexes[i];
      const movedFeature = editedData.features[i];
      updatedData = updatedData.replaceGeometry(selectedIndex, movedFeature.geometry);
    }
    return updatedData.getObject();
  };

  getScaleAction = (
    startDragPoint: Position,
    currentPoint: Position,
    editType: string,
    props: ModeProps<FeatureCollection>
  ) => {
    if (!this._selectedEditHandle || this._isSinglePointGeometrySelected()) {
      return null;
    }

    const oppositeHandle = this._getOppositeScaleHandle(this._selectedEditHandle);
    const origin = getCoord(oppositeHandle);

    const scaleFactor = getScaleFactor(
      origin,
      this._previousMouseCoords || startDragPoint,
      currentPoint
    );
    const scaledFeatures = turfTransformScale(this._selectedGeometry, scaleFactor, {
      origin
    });

    this._previousMouseCoords = currentPoint;

    return {
      updatedData: this._getUpdatedData(props, scaledFeatures),
      editType,
      editContext: {
        featureIndexes: props.selectedIndexes
      }
    };
  };

  updateCursor = (props: ModeProps<FeatureCollection>): boolean => {
    if (this._selectedEditHandle && this._selectedGeometry) {
      if (this._cursor) {
        props.onUpdateCursor(this._cursor);
        return true;
      }

      // Get resize cursor direction from the hovered scale editHandle (e.g. nesw or nwse)
      const centroid = turfCentroid(this._selectedGeometry);
      const bearing = turfBearing(centroid, this._selectedEditHandle);
      const positiveBearing = bearing < 0 ? bearing + 180 : bearing;
      if (
        (positiveBearing >= 0 && positiveBearing <= 90) ||
        (positiveBearing >= 180 && positiveBearing <= 270)
      ) {
        this._cursor = 'nesw-resize';
        props.onUpdateCursor('nesw-resize');
      } else {
        this._cursor = 'nwse-resize';
        props.onUpdateCursor('nwse-resize');
      }
      return true;
    }

    if (!this._isCompositionMode) {
      props.onUpdateCursor(null);
    }
    this._cursor = null;
    return false;
  };

  handlePointerMove(event: PointerMoveEvent, props: ModeProps<FeatureCollection>) {
    if (!this._isScaling) {
      const selectedEditHandle = getPickedEditHandle(event.picks);
      this._selectedEditHandle =
        selectedEditHandle && selectedEditHandle.properties.editHandleType === 'scale'
          ? selectedEditHandle
          : null;
    }

    if (!this._isCompositionMode) {
      this.updateCursor(props);
    }
  }

  handleStartDragging(event: StartDraggingEvent, props: ModeProps<FeatureCollection>) {
    if (this._selectedEditHandle) {
      this._isScaling = true;
    }
  }

  handleDragging(event: DraggingEvent, props: ModeProps<FeatureCollection>) {
    if (!this._isScaling) {
      return;
    }

    const scaleAction = this.getScaleAction(
      event.pointerDownMapCoords,
      event.mapCoords,
      'scaling',
      props
    );
    if (scaleAction) {
      props.onEdit(scaleAction);
    }

    event.cancelPan();
  }

  handleStopDragging(event: StopDraggingEvent, props: ModeProps<FeatureCollection>) {
    if (this._isScaling) {
      // Scale the geometry
      const scaleAction = this.getScaleAction(
        event.pointerDownMapCoords,
        event.mapCoords,
        'scaled',
        props
      );
      if (scaleAction) {
        props.onEdit(scaleAction);
      }

      this._selectedGeometry = null;
      this._selectedEditHandle = null;
      this._previousMouseCoords = null;
      this._cursor = null;
      this._isScaling = false;
    }
  }

  getGuides(props: ModeProps<FeatureCollection>) {
    this._cornerGuidePoints = [];
    this._selectedGeometry = this.getSelectedFeaturesAsFeatureCollection(props);

    // Add buffer to the enveloping box if a single Point feature is selected
    const featureWithBuffer = this._isSinglePointGeometrySelected()
      ? turfBuffer(this._selectedGeometry, 1)
      : this._selectedGeometry;

    const boundingBox = bboxPolygon(bbox(featureWithBuffer));
    const cornerGuidePoints = [];

    coordEach(boundingBox, (coord, coordIndex) => {
      if (coordIndex < 4) {
        // Get corner midpoint guides from the enveloping box
        const cornerPoint = point(coord, {
          guideType: 'editHandle',
          editHandleType: 'scale',
          index: coordIndex
        });
        cornerGuidePoints.push(cornerPoint);
      }
    });

    this._cornerGuidePoints = cornerGuidePoints;
    return featureCollection([boundingBox, ...this._cornerGuidePoints]);
  }
}

function getScaleFactor(centroid: Position, startDragPoint: Position, currentPoint: Position) {
  const startDistance = turfDistance(centroid, startDragPoint);
  const endDistance = turfDistance(centroid, currentPoint);
  return endDistance / startDistance;
}
