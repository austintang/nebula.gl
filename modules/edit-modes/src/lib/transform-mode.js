// @flow

import { featureCollection } from '@turf/helpers';
import { getPickedEditHandle } from '../utils';
import type {
  PointerMoveEvent,
  StartDraggingEvent,
  StopDraggingEvent,
  DraggingEvent,
  ModeProps
} from '../types.js';
import type { FeatureCollection } from '../geojson-types.js';
import { BaseGeoJsonEditMode } from './geojson-edit-mode';
import { TranslateMode } from './translate-mode';
import { ScaleMode } from './scale-mode';
import { RotateMode } from './rotate-mode';

export class TransformMode extends BaseGeoJsonEditMode {
  _transformMode: ?string;
  _editModes: Object = {
    scale: new ScaleMode({ isCompositionMode: true }),
    rotate: new RotateMode({ isCompositionMode: true }),
    translate: new TranslateMode({ isCompositionMode: true })
  };

  getGuides(props: ModeProps<FeatureCollection>) {
    const { rotate, scale } = this._editModes;
    const scaleGuides = rotate.getIsRotating() ? [] : scale.getGuides(props).features;
    const rotateGuides = rotate
      .getGuides(props)
      .features.filter(feature => feature.geometry.type !== 'Polygon');

    return featureCollection([...scaleGuides, ...rotateGuides]);
  }

  handlePointerMove(event: PointerMoveEvent, props: ModeProps<FeatureCollection>) {
    let isCursorUpdated = false;
    Object.values(this._editModes).forEach(mode => {
      if (!mode) {
        return;
      }

      if (typeof mode.handlePointerMove === 'function') {
        mode.handlePointerMove(event, props);
      }
      if (typeof mode.updateCursor === 'function') {
        if (!isCursorUpdated) {
          isCursorUpdated = mode.updateCursor(props);
        }
      }
    });

    if (!isCursorUpdated) {
      props.onUpdateCursor(null);
    }
  }

  handleStartDragging(event: StartDraggingEvent, props: ModeProps<FeatureCollection>) {
    const { picks, pointerDownPicks } = event;
    const editHandle = getPickedEditHandle(picks);
    if (editHandle) {
      this._transformMode = editHandle.properties.editHandleType;
    } else if (this.isSelectionPicked(pointerDownPicks || picks, props)) {
      this._transformMode = 'translate';
    } else {
      this._transformMode = null;
    }

    const transformMode = this._editModes[this._transformMode];
    if (transformMode) {
      transformMode.handleStartDragging(event, props);
    }
  }

  handleDragging(event: DraggingEvent, props: ModeProps<FeatureCollection>) {
    const transformMode = this._editModes[this._transformMode || ''];
    if (transformMode) {
      transformMode.handleDragging(event, props);
    }
  }

  handleStopDragging(event: StopDraggingEvent, props: ModeProps<FeatureCollection>) {
    const transformMode = this._editModes[this._transformMode || ''];
    if (transformMode) {
      transformMode.handleStopDragging(event, props);
    }
    this._transformMode = null;
  }
}
