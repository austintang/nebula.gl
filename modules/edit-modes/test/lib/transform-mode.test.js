// @flow

import { TransformMode } from '../../src/lib/transform-mode';
import {
  createFeatureCollectionProps,
  createPointerMoveEvent,
  createStartDraggingEvent,
  createStopDraggingEvent
} from '../test-utils.js';
import { type Pick } from '../../src/types';

let transformMode: TransformMode;

let warnBefore;
beforeEach(() => {
  warnBefore = console.warn; // eslint-disable-line
  // $FlowFixMe
  console.warn = function() {}; // eslint-disable-line
  transformMode = new TransformMode();
});

afterEach(() => {
  // $FlowFixMe
  console.warn = warnBefore; // eslint-disable-line
});

describe('Transform - translate mode', () => {
  const mockMove = (picks: Pick[], props: Object) => {
    const moveEvent = createPointerMoveEvent([-1, -1], picks);
    transformMode.handlePointerMove(moveEvent, props);

    const startDragEvent = createStartDraggingEvent([-1, -1], [-1, -1], picks);
    transformMode.handleStartDragging(startDragEvent, props);

    const stopDragEvent = createStopDraggingEvent([2, 3], [-1, -1], picks);
    transformMode.handleStopDragging(stopDragEvent, props);
  };

  test('Selected polygon feature can be translated', () => {
    const mockOnEdit = jest.fn();
    const props = createFeatureCollectionProps({
      selectedIndexes: [2],
      onEdit: mockOnEdit
    });
    mockMove([{ index: 2, isGuide: false, object: null }], props);

    expect(mockOnEdit).toHaveBeenCalledTimes(1);
    const movedFeature = mockOnEdit.mock.calls[0][0].updatedData.features[2];
    expect(movedFeature).toMatchSnapshot();
    expect(props.data.features[2]).not.toEqual(movedFeature);
  });

  test('Non-picked selected polygon feature cannnot be translated', () => {
    const mockOnEdit = jest.fn();
    const props = createFeatureCollectionProps({
      selectedIndexes: [2],
      onEdit: mockOnEdit
    });
    mockMove([], props);
    expect(mockOnEdit).toHaveBeenCalledTimes(0);
  });

  test('Picked non-selected polygon feature cannnot be translated', () => {
    const mockOnEdit = jest.fn();
    const props = createFeatureCollectionProps({
      selectedIndexes: [0],
      onEdit: mockOnEdit
    });
    mockMove([{ index: 2, isGuide: false, object: null }], props);
    expect(mockOnEdit).toHaveBeenCalledTimes(0);
  });
});

describe('Transform - scale mode', () => {
  const mockScale = (picks: Pick[], props: Object) => {
    transformMode.getGuides(props);

    const moveEvent = createPointerMoveEvent([-1, -1], picks);
    transformMode.handlePointerMove(moveEvent, props);

    const startDragEvent = createStartDraggingEvent([-1, -1], [-1, -1], picks);
    transformMode.handleStartDragging(startDragEvent, props);

    const stopDragEvent = createStopDraggingEvent([2, 3], [-1, -1], picks);
    transformMode.handleStopDragging(stopDragEvent, props);
  };

  test('Selected polygon feature can be scaled', () => {
    const mockOnEdit = jest.fn();
    const props = createFeatureCollectionProps({
      selectedIndexes: [2],
      onEdit: mockOnEdit
    });
    const mockPicks = [
      {
        index: 2,
        isGuide: true,
        object: {
          type: 'Feature',
          geometry: { type: 'Point', coordinates: [-2, -2] },
          properties: { guideType: 'editHandle', editHandleType: 'scale', index: 0 }
        }
      }
    ];
    mockScale(mockPicks, props);

    expect(mockOnEdit).toHaveBeenCalledTimes(1);
    const scaledFeature = mockOnEdit.mock.calls[0][0].updatedData.features[2];
    expect(scaledFeature).toMatchSnapshot();
    expect(props.data.features[2]).not.toEqual(scaledFeature);
  });

  test('Selected polygon feature without edit handle picks cannot be scaled', () => {
    const mockOnEdit = jest.fn();
    const props = createFeatureCollectionProps({
      selectedIndexes: [2],
      onEdit: mockOnEdit
    });

    mockScale([], props);
    expect(mockOnEdit).toHaveBeenCalledTimes(0);
  });
});

describe('Transform - rotate mode', () => {
  const mockRotate = (picks: Pick[], props: Object) => {
    transformMode.getGuides(props);

    const moveEvent = createPointerMoveEvent([-1, -1], picks);
    transformMode.handlePointerMove(moveEvent, props);

    const startDragEvent = createStartDraggingEvent([-1, -1], [-1, -1], picks);
    transformMode.handleStartDragging(startDragEvent, props);

    const stopDragEvent = createStopDraggingEvent([2, 3], [-1, -1], picks);
    transformMode.handleStopDragging(stopDragEvent, props);
  };

  test('Selected polygon feature can be rotated', () => {
    const mockOnEdit = jest.fn();
    const props = createFeatureCollectionProps({
      selectedIndexes: [2],
      onEdit: mockOnEdit
    });
    const mockPicks = [
      {
        index: 2,
        isGuide: true,
        object: {
          type: 'Feature',
          geometry: { type: 'Point', coordinates: [-2, -2] },
          properties: { guideType: 'editHandle', editHandleType: 'rotate' }
        }
      }
    ];
    mockRotate(mockPicks, props);

    expect(mockOnEdit).toHaveBeenCalledTimes(1);
    const scaledFeature = mockOnEdit.mock.calls[0][0].updatedData.features[2];
    expect(scaledFeature).toMatchSnapshot();
    expect(props.data.features[2]).not.toEqual(scaledFeature);
  });

  test('Selected polygon feature without edit handle picks cannot be rotated', () => {
    const mockOnEdit = jest.fn();
    const props = createFeatureCollectionProps({
      selectedIndexes: [2],
      onEdit: mockOnEdit
    });

    mockRotate([], props);
    expect(mockOnEdit).toHaveBeenCalledTimes(0);
  });
});
