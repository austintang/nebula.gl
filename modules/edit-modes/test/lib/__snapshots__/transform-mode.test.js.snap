// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Transform - rotate mode Selected polygon feature can be rotated 1`] = `
Object {
  "geometry": Object {
    "coordinates": Array [
      Array [
        Array [
          0.5272280648573542,
          0.8431285896095478,
        ],
        Array [
          -1.437857681943001,
          0.47152591660008025,
        ],
        Array [
          -1.0663539798351849,
          -1.4936478169257912,
        ],
        Array [
          0.8989219206248436,
          -1.1220394838939336,
        ],
        Array [
          0.5272280648573542,
          0.8431285896095478,
        ],
      ],
      Array [
        Array [
          0.12886514525962411,
          0.2589380256236096,
        ],
        Array [
          0.3146948902101485,
          -0.7236467186215011,
        ],
        Array [
          -0.6679218708436565,
          -0.9094572530721381,
        ],
        Array [
          -0.8537040799629949,
          0.07312890620550877,
        ],
        Array [
          0.12886514525962411,
          0.2589380256236096,
        ],
      ],
    ],
    "type": "Polygon",
  },
  "properties": Object {},
  "type": "Feature",
}
`;

exports[`Transform - scale mode Selected polygon feature can be scaled 1`] = `
Object {
  "geometry": Object {
    "coordinates": Array [
      Array [
        Array [
          -0.5809507320924467,
          -0.5809702731526293,
        ],
        Array [
          1,
          -0.5809702731526291,
        ],
        Array [
          1,
          1,
        ],
        Array [
          -0.5809702731526158,
          1,
        ],
        Array [
          -0.5809507320924467,
          -0.5809702731526293,
        ],
      ],
      Array [
        Array [
          -0.1857336510214509,
          -0.18572770486447193,
        ],
        Array [
          -0.18574097865473504,
          0.6047574317118429,
        ],
        Array [
          0.6047530071150504,
          0.6047574317118425,
        ],
        Array [
          0.6047554496594785,
          -0.18572770486447193,
        ],
        Array [
          -0.1857336510214509,
          -0.18572770486447193,
        ],
      ],
    ],
    "type": "Polygon",
  },
  "properties": Object {},
  "type": "Feature",
}
`;

exports[`Transform - translate mode Selected polygon feature can be translated 1`] = `
Object {
  "geometry": Object {
    "coordinates": Array [
      Array [
        Array [
          1.9993896804101041,
          3.0004573272249484,
        ],
        Array [
          3.999389680410104,
          3.0004573272249484,
        ],
        Array [
          4.003051832614915,
          5.0004573272249475,
        ],
        Array [
          2.003051832614915,
          5.0004573272249475,
        ],
        Array [
          1.9993896804101041,
          3.0004573272249484,
        ],
      ],
      Array [
        Array [
          2.4999614397826235,
          3.5004573272249484,
        ],
        Array [
          2.501792166258838,
          4.500457327224948,
        ],
        Array [
          3.501792166258838,
          4.500457327224948,
        ],
        Array [
          3.4999614397826235,
          3.5004573272249484,
        ],
        Array [
          2.4999614397826235,
          3.5004573272249484,
        ],
      ],
    ],
    "type": "Polygon",
  },
  "properties": Object {},
  "type": "Feature",
}
`;
